### ISSUE TITLE
groot: low CPU utilization, from vars decoded in the main `rtree.Reader.Read` goroutine

### ISSUE THREAD

**Issue by bburghgr** [go-hep/hep#1020]
When running a trivial event loop (i.e. just `return nil` for every event, without doing any real work), I've noticed that my CPU utilization is fairly low (only about 230% on average, on a 4 core / 8 hyperthread machine). It seems that the I/O workers are blocked waiting for the event loop to progress, despite there being the minimum possible amount of work for the event loop to do.

I'm guessing this may be due to the way the `rtree.Reader.Read` decodes values from their underlying `[]byte` buffers on demand (and copies them to the location of their associated `ReadVar.Value` in memory). Decoding from the underlying `[]byte` happens at the start of each event, within the goroutine that runs the function supplied to `Read`. This can apparently be a non-negligible amount of work compared to decompression (the `BigEndian` calls inside `RBuffer.ReadF32` account for about 20% of the CPU time in my test, decompression is about 50%, so they seem to be in the same ballpark).

Perhaps it would make sense to decode the vars from the `[]byte` into a slice of the appropriate value, e.g. a `[]float32`, from within the `blkreader.run` workers? That would reduce the work done in the main `rtree.Reader.Read` goroutine to a simple copy from e.g. some offset of a `[]float32` into the location pointed to by a `*float32`. Though it may just be moving a cache miss from the `[]byte` decoding and into the copy operation, so I'm not certain it would really be an improvement.

I haven't tried to prototype a workaround, but quickly hacking the `RBuffer.ReadF32` function to do no useful work (`return 0` before the `BigEndian` conversion), I'm able to speed up the event loop and get somewhere around 350% CPU utilization (and the file finishes in ~20s wall time, instead of 30s). Therefore, I'm reasonably confident that this is the main bottleneck when it comes to CPU utilization (assuming the user code take negligible time per event, due to e.g. low selection efficiency, or if the user writes their code in a way that can safely process events in parallel). I just thought I should mention it, since it was a bit surprising that such a seemingly trivial operation would be the bottleneck.

**Comment by bburghgr** [comment]
I've looked into this a bit more. For reference, I'm using a test file with 1M events, 10k branches of float32 data, and running on a 4 core CPU (hypterthreading enabled).

Long story short, I'm pretty confident that the CPU time spent in `RBuffer.ReadF32` is really from memory access, i.e. the large number of branches is enough to evict the underlying buffers from my CPU cache between events. This means, even if the decoding was moved into the `kbreader.run` workers, the main goroutine would still experience most of the slowdown (when it tries to access the values from the `[]float32` and copy them to the right location in memory). So, for my specific benchmark, I expect that moving the decoding step into the workers would lead to a negligible improvement. Other workflows may see something different.

I'm still running some tests, but the short version is that column-wise access (processing 1 `ReadVar` at a time, in sequence, from separate `rtree.Reader`s) seems to minimize the total CPU time, and reduces the time spent in `RBuffer.ReadF32` to something negligible. I'll try to post more details when I'm done running the tests, in case it illuminates anything.

**Comment by sbinet** [comment]
using a "tower" of builtins from which `ReadVar`s would subslice accordingly sounds a lot like what `rntup` (or `uproot` and its `Cursor` interface) are supposed to do.

I put that on the back burner for 2 reasons:
1) I was lazy, and it seemed to me I would have to implement polymorphic `rbasket`s, parallel to what I had to do for `Leaf` and `rleaf` (and generics weren't available yet)
2) I couldn't find a way to robustly and user-friendly make sure subslices wouldn't be used after an event/row had been processed (hence the copy into "user context"). perhaps this could have been done with a "golint/govet" code checker...

`1)` could be addressed, it's "just" code.
`2)` is a bit more involved. perhaps it will become moot once/when `rntup` is definitely released with a `v1` (and groot user(s) start to want to read/write those things)

**Comment by bburghgr** [comment]
That makes sense. For issue 2, if you're willing to allocate new memory for each basket, then I suppose there's no harm in letting the user retain a subslice, but that may slow things down a bit to support something that *seems* like bad practice. I can see why that may not be worth the tradeoff. In any case, you're right, the `RNTuple` may change the situation (and it's probably not worth it to worry about `rtree` performance *too* much, given that already beating `ROOT::TTree` performance, and it may become a legacy format in the somewhat near future).

I'm not sure there's any action worth taking here, so I'll go ahead and close this issue for now.

In case anyone comes across this later and is interested, just for completeness, here are the runtimes for some of the scenarios I've tested on my machine, using the test file described above:

Running "normally": about 75s CPU time (30s wall time).

Running with `GOMAXPROCS=1`: about 50s CPU time. Decompression takes about 25s, and about 15s are spent on the `binary.BigEndian.Uint32` call from within `RBuffer.ReadF32`.

Running with `GOMAXPROCS=1` and reading 1 `ReadVar` at a time (so 10k separate event loops): about 36s CPU time, 23s spent on decompression, and just over 1s in `RBuffer.ReadF32`. Just to give some sense for how much time is spent accessing memory if there are enough branches being read for cache misses to occur.

Running spans of 8 `ReadVar`s at a time, I manage to get about 430% CPU usage, with almost exactly 60s CPU time and just under 14s wall time. This is the fastest wall time I've managed for iterating through the full contents of this file.

Using `uproot` in python takes about 20s CPU time (25s if `cProfile` is enabled, of which 15s are spent on decompression). This is actually calling the function to convert a tree into a numpy array, so it may be doing slightly more work. This is the fastest CPU time I've managed for the test file.

The `ROOT::RDataFrame` has an `AsNumpy()` method similar to the above, which takes about 180s.

Filling a histogram per branch with `TTree::Draw` takes 60s. I expect this to approximate the minimum CPU cost (and maximum single-threaded throughput) that `ROOT` can achieve.

In `pyROOT`, naive use of `TTree::GetEntry()` takes 65 CPU seconds, if the branches are not touched from the python side. Reading from the branches the obvious way (e.g. per event with attribute syntax) takes... too long, events process at something like 10 Hz. This is probably the easiest way to read from a tree in the official `ROOT` distribution, so it's unfortunate that this is also (by far) the slowest.

All of these are with a trivial workload in the event loop, and make sense for different workflows (if ever), so it's probably not representative of real-world use.
